Sources/protoc-gen-swift/MessageGenerator.swift

```diff
@@ -739,3 +944,40 @@ class MessageGenerator {
         }
     }
 }
+
+// The logic for this check comes from google/protobuf; the C++ and Java generators specificly.
+//
+// This is a helper for generating isInitialized methods.
+fileprivate func HasRequiredFields(_ fieldDesc: Google_Protobuf_FieldDescriptorProto, context: Context) -> Bool {
+  assert(fieldDesc.isGroup || fieldDesc.isMessage)
+  var alreadySeen = Set<Google_Protobuf_DescriptorProto>()
+
+  func HasRequiredFieldsInner(_ msgDesc: Google_Protobuf_DescriptorProto) -> Bool {
+    if alreadySeen.contains(msgDesc) {
+      // First required thing found causes this to return true, so one can assume if it is already
+      // visited, it didn't have required fields.
+      return false
+    }
+    alreadySeen.insert(msgDesc)
+
+    // If it can support extesions, and extension could be a message with required fields.
+    if msgDesc.extensionRange.count > 0 {
+      return true
+    }
+
+    for f in msgDesc.field {
+      if f.label == .required {
+        return true
+      }
+      if (f.isMessage || f.isGroup) &&
+        HasRequiredFieldsInner(context.getMessageForPath(path: f.typeName)!) {
+        return true
+      }
+    }
+
+    return false
+  }
+
+  let msgDesc = context.getMessageForPath(path: fieldDesc.typeName)!
```

Comment:
Our use of getMessageForPath suddenly worries me.  [descriptor.proto's](https://github.com/google/protobuf/blob/master/src/google/protobuf/descriptor.proto#L174-L178) seems to say that string we're using might not be full paths, and could just be the leaf from the parent, but the way we use it it seems like it always has to be full.